server/server/src/billing/app_store_service.rs

```diff
@@ -0,0 +1,131 @@
+            )
+        })?;
+
+    let latest_receipt = validated_receipts
+        .get(0)
+        .ok_or(ClientError(UpgradeAccountAppStoreError::InvalidAuthDetails))?;
+    let expires_date_ms: i64 = latest_receipt
+        .expires_date_ms
+        .parse()
+        .map_err::<ServerError<UpgradeAccountAppStoreError>, _>(|err| {
+            internal!(
+                "Cannot parse expires_date_ms into i64, latest_receipt: {:?}, err: {:?}",
+                latest_receipt,
+                err
+            )
+        })?;
+
+    if latest_receipt.app_account_token != app_account_token
+        || latest_receipt.original_transaction_id != original_transaction_id
+        || expires_date_ms < get_time().0
+    {
+        return Err(ClientError(UpgradeAccountAppStoreError::InvalidAuthDetails));
+    }
+
+    Ok(expires_date_ms as UnixTimeMillis)
+}
+
+pub fn decode_verify_notification(
+    config: &AppleConfig, request_body: &Bytes,
+) -> Result<NotificationResponseBody, ServerError<AppStoreNotificationError>> {
+    let encoded_resp: EncodedNotificationResponseBody = serde_json::from_slice(request_body)
+        .map_err(|_| ClientError(AppStoreNotificationError::InvalidJWS))?;
+
+    validate_jwt(config, &encoded_resp.signed_payload)
+}
+
+fn validate_jwt<T: Serialize + DeserializeOwned>(
+    config: &AppleConfig, token: &str,
+) -> Result<T, ServerError<AppStoreNotificationError>> {
+    let header = decode_header(token)?;
+    let cert_chain: Vec<Vec<u8>> = header
+        .x5c
+        .ok_or(ClientError(AppStoreNotificationError::InvalidJWS))?
+        .into_iter()
+        .map(|cert| base64::decode(cert.as_bytes()))
+        .collect::<Result<Vec<Vec<u8>>, base64::DecodeError>>()?;
+
+    let certs: Vec<X509Certificate> = cert_chain
+        .iter()
+        .map(|cert| parse_x509_certificate(cert.as_slice()).map(|(_, cert)| cert))
+        .collect::<Result<Vec<X509Certificate>, x509_parser::nom::Err<X509Error>>>()
+        .map_err(|_| ClientError(AppStoreNotificationError::InvalidJWS))?;
+
+    for i in 0..certs.len() {
+        if i != certs.len() - 1 {
+            certs[i]
+                .verify_signature(Some(&certs[i + 1].subject_pki))
+                .map_err(|_| ClientError(AppStoreNotificationError::InvalidJWS))?;
+        } else {
+            certs[i]
+                .verify_signature(Some(
+                    &parse_x509_certificate(config.apple_root_cert.as_slice())
+                        .map_err(|err| internal!("{:?}", err))?
+                        .1
+                        .subject_pki,
+                ))
+                .map_err(|_| ClientError(AppStoreNotificationError::InvalidJWS))?;
+        }
+    }
+
+    let pem = format!(
+        "-----BEGIN PUBLIC KEY-----\n{}\n-----END PUBLIC KEY-----",
+        base64::encode(certs[0].public_key().raw)
+    );
+    let key = DecodingKey::from_ec_pem(pem.as_bytes())?;
```

Comment:
Since this wasn't the main issue, couldn't we use the non-pem one and not turn our `der` into `pem` ourselves?

Reply:
I tried to use der and it did not work as expected, likely a little thing I am missing.